spring-integration-core/src/main/java/org/springframework/integration/handler/advice/RequestHandlerRateLimiterAdvice.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.handler.advice;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLongArray;
+
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessagingException;
+import org.springframework.util.Assert;
+
+/**
+ * Rate limiter advice, which will ensure call to service is Rate limited within set rate limit.
+ * Request calls breaching the rate limit will wait for time rate limit is not breached.
+ *
+ * @author Meherzad Lahewala
+ * @since 5.1
+ */
+public class RequestHandlerRateLimiterAdvice extends AbstractRequestHandlerAdvice {
+
+	private final AtomicInteger currentIndex;
+	private final long maxDelayInMillis;
+	private final long oneTimeUnitInMillis;
+	private final AtomicLongArray queue;
+	private final int rate;
+
+	/**
+	 * Create an instance of RequestHandlerRateLimiterAdvice with the provided arguments.
+	 *
```

Comment:
We prefer do not have blank lines in the method Javadocs: https://github.com/spring-projects/spring-framework/wiki/Code-Style#javadoc-formatting